broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/SnapshotDirectorPartitionTransitionStep.java

```diff
@@ -0,0 +1,110 @@
+import io.camunda.zeebe.util.sched.future.ActorFuture;
+import io.camunda.zeebe.util.sched.future.CompletableActorFuture;
+import java.time.Duration;
+
+public class SnapshotDirectorPartitionTransitionStep implements PartitionTransitionStep {
+
+  @Override
+  public ActorFuture<Void> prepareTransition(
+      final PartitionTransitionContext context, final long term, final Role targetRole) {
+    final AsyncSnapshotDirector snapshotDirector = context.getSnapshotDirector();
+    if (snapshotDirector != null
+        && (shouldInstallOnTransition(targetRole, context.getCurrentRole())
+            || targetRole == Role.INACTIVE)) {
+      final var director = context.getSnapshotDirector();
+      context.getComponentHealthMonitor().removeComponent(director.getName());
+      context.getRaftPartition().getServer().removeCommittedEntryListener(director);
+      final ActorFuture<Void> future = director.closeAsync();
+      future.onComplete(
+          (ok, error) -> {
+            if (error == null) {
+              context.setSnapshotDirector(null);
+            }
+          });
+      return future;
+    } else {
+      return CompletableActorFuture.completed(null);
+    }
+  }
+
+  @Override
+  public ActorFuture<Void> transitionTo(
+      final PartitionTransitionContext context, final long term, final Role targetRole) {
+    if ((context.getSnapshotDirector() == null && targetRole != Role.INACTIVE)
+        || shouldInstallOnTransition(targetRole, context.getCurrentRole())) {
+      final var server = context.getRaftPartition().getServer();
+
+      final Duration snapshotPeriod = context.getBrokerCfg().getData().getSnapshotPeriod();
+      final AsyncSnapshotDirector director;
+      if (targetRole == Role.LEADER) {
+        director = createSnapshotDirectorOfLeader(context, server, snapshotPeriod);
+      } else {
+        director = createSnapshotDirectorOfFollower(context, snapshotPeriod);
+      }
+
+      final var future = context.getActorSchedulingService().submitActor(director);
+      future.onComplete(
+          (ok, error) -> {
+            if (error == null) {
+              context.setSnapshotDirector(director);
+              context.getComponentHealthMonitor().registerComponent(director.getName(), director);
+            }
+          });
+      return future;
+
+    } else {
+      return CompletableActorFuture.completed(null);
+    }
+  }
+
+  @Override
+  public String getName() {
+    return "AsyncSnapshotDirector";
+  }
+
+  private AsyncSnapshotDirector createSnapshotDirectorOfLeader(
+      final PartitionTransitionContext context,
+      final RaftPartitionServer server,
+      final Duration snapshotPeriod) {
+    final var director =
+        AsyncSnapshotDirector.ofProcessingMode(
+            context.getNodeId(),
+            context.getPartitionId(),
+            context.getStreamProcessor(),
+            context.getStateController(),
+            snapshotPeriod);
+
+    server.addCommittedEntryListener(director);
```

Comment:
I would prefer this line to also be part of the code block
```
      final var future = context.getActorSchedulingService().submitActor(director);
      future.onComplete(
          (ok, error) -> {
            if (error == null) {
              context.setSnapshotDirector(director);
              context.getComponentHealthMonitor().registerComponent(director.getName(), director);
            }
          });
```

Unless there is a reason why this is done here, of course.

Reply:
This is only done when it is a leader.

Reply:
Still, we could check for the role and do it only execute it when we transition to leader.
My concern is that if we have an error, then we have registered an entry listener and no longer have a reference to it.